sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -1043,6 +1045,34 @@ object TypeCoercion {
       }
     }
   }
+
+  /**
+   * A special rule to support string literal as the second argument of date_add/date_sub functions,
+   * to keep backward compatibility as a temporary workaround.
+   * TODO(SPARK-28589): implement ANSI type type coercion and handle string literals.
+   */
+  object StringLiteralCoercion extends TypeCoercionRule {
```

Comment:
This PR seems to have a side effect to change the behavior on a general arithmetic (`1 + date`) instead of `date_add/date_sub` function. Apache Spark 2.4.x doesn't allow those artithemetic.